text/0000-bench-utils.md

```diff
@@ -0,0 +1,138 @@
+- Feature Name: black_box-and-clobber
+- Start Date: 2018-03-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds two functions to `core::mem`: `black_box` and `clobber`, which are
+mainly useful for writing benchmarks.
+
+# Motivation
+[motivation]: #motivation
+
+The `black_box` and `clobber` functions are useful for writing synthetic
+benchmarks where, due to the constrained nature of the benchmark, the compiler
+is able to perform optimizations that wouldn't otherwise trigger in practice.
+
+The implementation of these functions is backend-specific and requires inline
+assembly. Such that if the standard library does not provide them, the users are
+required to use brittle workarounds on nightly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `mem::black_box`
+
+The function:
+
+```rust
+pub fn black_box<T>(x: T) -> T;
+```
+
+prevents the value `x` from being optimized away and flushes pending reads/writes
+to memory. It does not prevent optimizations on the expression generating the
+value `x` nor on the return value of the function. For
+example ([`rust.godbolt.org`](https://godbolt.org/g/YP2GCJ)):
+
+```rust
+fn foo(x: i32) -> i32{ 
+  mem::black_box(2 + x);
+  3
+}
+let a = foo(2);
+```
+
+Here, the compiler can simplify the expression `2 + x` into `2 + 2` and then
+`4`, but it is not allowed to discard `4`. Instead, it must store `4` into a
+register even though it is not used by anything afterwards.
```

Comment:
Nit, but this doesn't really match what the implementation of black_box does (and what the linked godbolt shows). It forces the value into memory, and any register traffic is due to that (e.g. in the linked godbolt, the `add edi, 2` is not a dead definition, it's immediately stored to `[rbp - 4]`, and *that store* is dead).

Reply:
@rkruppe how would you formulate this for the guide-level explanation? Maybe I can just say that `4` must be stored into memory and leave it at that.

Reply:
That seems decent. But it's tricky to say *anything* about this sort of feature, because it's inherently very tied to optimizer capabilities and the resulting machine code =/

Reply:
So I've reworded this part a bit, but maybe for the guide level explanation we might want to be even more vague and just leave it at: "the value cannot be discarded" or "prevents the value `x` from being optimized away".